File: backend/app/api/schemas/devices.py
"""DTOs de dispositivos: registro, lectura y resultados de escaneo."""

from __future__ import annotations

from pydantic import BaseModel, Field

from backend.app.domain.devices.models import (
    Device,
    DeviceCapabilities,
    DeviceType,
    DiscoveredDevice,
)


class DeviceCapabilitiesRead(BaseModel):
    """Que sabe hacer el dispositivo. Es lo que decide que renderiza la UI.

    Los campos se enumeran uno a uno en vez de volcar el modelo de dominio: esto
    es el contrato publicado y añadir un campo al dominio no debe cambiarlo por
    accidente (ni al reves).
    """

    rgb: bool
    brightness: bool
    effects: bool
    addressable: bool
    segments: bool
    white_channel: bool
    music_mode: bool

    @classmethod
    def from_domain(cls, capabilities: DeviceCapabilities) -> DeviceCapabilitiesRead:
        return cls(
            rgb=capabilities.rgb,
            brightness=capabilities.brightness,
            effects=capabilities.effects,
            addressable=capabilities.addressable,
            segments=capabilities.segments,
            white_channel=capabilities.white_channel,
            music_mode=capabilities.music_mode,
        )


class DeviceRead(BaseModel):
    """Un dispositivo registrado, con su estado de enlace.

    `connected` NO viene del dominio `Device` (que solo modela identidad
    persistida) sino del estado global: se compone aqui, en la frontera, porque
    es justo lo que un cliente necesita ver junto.
    """

    #: Cadena, no UUID: el JSON no tiene tipo UUID y el frontend no debe
    #: inventarse la conversion.
    id: str
    name: str
    adapter_type: DeviceType
    address: str | None = None
    enabled: bool
    auto_connect: bool
    connected: bool
    capabilities: DeviceCapabilitiesRead

    @classmethod
    def from_domain(cls, device: Device, *, connected: bool) -> DeviceRead:
        return cls(
            id=str(device.id),
            name=device.name,
            adapter_type=device.adapter_type,
            address=device.address,
            enabled=device.enabled,
            auto_connect=device.auto_connect,
            connected=connected,
            capabilities=DeviceCapabilitiesRead.from_domain(device.capabilities),
        )


class DiscoveredDeviceRead(BaseModel):
    """Resultado de `GET /devices/scan` (README 16).

    Sin `id` a proposito: lo descubierto todavia no existe en la base. Lo que le
    da identidad es `POST /devices`.
    """

    name: str | None = None
    address: str
    rssi: int | None = None

    @classmethod
    def from_domain(cls, discovered: DiscoveredDevice) -> DiscoveredDeviceRead:
        return cls(name=discovered.name, address=discovered.address, rssi=discovered.rssi)


class DeviceCreateRequest(BaseModel):
    """Cuerpo de `POST /devices`.

    Este endpoint no esta en el README y hace falta: `/devices/{id}/connect`
    presupone un id persistido y `/devices/scan` devuelve resultados efimeros
    sin id. Registrar lo descubierto es lo que evita escribir una direccion a
    mano en el codigo (NEXT_STEPS 4.3).

    `adapter_type` es opcional porque el proceso construye UN adaptador
    (`LED_ROOM_DEVICE_ADAPTER`) y el escaneo que produjo la direccion lo hizo el
    descubridor de esa misma familia. Se acepta para que el cliente pueda
    afirmar lo que cree estar registrando, y se rechaza con 422 si contradice a
    la configuracion: aceptarlo y guardar otra cosa seria etiquetar mal el
    dispositivo en silencio.
    """

    name: str | None = Field(default=None, min_length=1, max_length=120)
    address: str = Field(min_length=1, max_length=255)
    adapter_type: DeviceType | None = None

    def to_domain(self) -> DiscoveredDevice:
        """Lo convierte en la entrada que espera el caso de uso de registro."""
        return DiscoveredDevice(name=self.name, address=self.address)

File: frontend/src/domain/devices.ts
/**
 * Dispositivos en el dominio del cliente.
 *
 * Los tipos del backend se conservan separados a proposito
 * (backend/app/domain/devices/models.py):
 *
 *   - `Device`        identidad registrada + lo que sabe hacer.
 *   - `DeviceLink`    estado efimero del enlace (conectado, rssi, ultimo error).
 *
 *   - `DiscoveredDevice`  resultado de un escaneo: **todavia no existe** en el
 *                         servidor, solo se ha visto por la radio.
 *
 * Colapsarlos fue justo el error que el backend corrigio: la UI pinta cosas
 * distintas con cada uno. Un `DiscoveredDevice` no tiene `id` porque no esta
 * registrado; darle uno invitaria a intentar conectarlo antes de registrarlo.
 */

/**
 * Que sabe hacer el dispositivo conectado.
 *
 * **Es lo unico que decide que controles se habilitan.** Nunca `adapterType`:
 * el tipo de adaptador es una decision de infraestructura y atarle la UI
 * obligaria a tocar componentes cada vez que se soporte hardware nuevo
 * (NEXT_STEPS A7).
 */
export interface DeviceCapabilities {
  readonly rgb: boolean
  readonly brightness: boolean
  readonly effects: boolean
  readonly addressable: boolean
  readonly segments: boolean
  readonly whiteChannel: boolean
  readonly musicMode: boolean
}

/** Nombre de una capacidad. Permite pedir "la capacidad X" sin cadenas sueltas. */
export type CapabilityName = keyof DeviceCapabilities

/** Un dispositivo registrado en el backend. */
export interface Device {
  readonly id: string
  readonly name: string
  /** Familia de hardware. Informativo: NO se usa para decidir la UI. */
  readonly adapterType: string
  readonly address: string | null
  readonly enabled: boolean
  readonly autoConnect: boolean
  readonly connected: boolean
  readonly capabilities: DeviceCapabilities
}

/** Estado del unico enlace que mantiene el proceso backend. */
export interface DeviceLink {
  readonly deviceId: string
  readonly connected: boolean
  readonly rssi: number | null
  /** Ultimo fallo del enlace. Se muestra: un fallo oculto es un fallo repetido. */
  readonly lastError: string | null
}

/**
 * Capacidades de un dispositivo que todavia no conocemos.
 *
 * Todo en `false`: sin dispositivo conectado no hay nada que se pueda hacer, y
 * asumir lo contrario dejaria controles habilitados que fallarian con 409.
 */
export const NO_CAPABILITIES: DeviceCapabilities = {
  rgb: false,
  brightness: false,
  effects: false,
  addressable: false,
  segments: false,
  whiteChannel: false,
  musicMode: false,
}

/**
 * Un dispositivo visto durante un escaneo BLE y **aun no registrado**.
 *
 * `name` es opcional en el anuncio BLE: hay tiras que no publican ninguno. El
 * `rssi` es la unica pista de cercania y por eso se conserva, aunque sea
 * ruidoso; `address` no es portable entre hosts (MAC en Linux, GUID en WinRT),
 * asi que sirve para registrar, no para que el usuario reconozca su tira.
 */
export interface DiscoveredDevice {
  readonly name: string | null
  readonly address: string
  readonly rssi: number | null
}

/**
 * Alta de un dispositivo en el servidor.
 *
 * No lleva `adapterType`: la familia de hardware la decide el backend, que es
 * quien sabe con que adaptador esta corriendo. Elegirla desde la UI seria
 * adivinar. El alta es **idempotente por direccion**: repetirla devuelve el
 * mismo dispositivo, nunca un duplicado ni un 409.
 */
export interface NewDevice {
  readonly name: string | null
  readonly address: string
}
